plat/rpi/rpi3/aarch64/plat_helpers.S

```diff
@@ -62,30 +62,31 @@ func plat_is_my_cpu_primary
 endfunc plat_is_my_cpu_primary
 
 	/* -----------------------------------------------------
-	 * void plat_secondary_cold_boot_setup (void);
+	 * void plat_wait_for_warm_boot (void);
 	 *
 	 * This function performs any platform specific actions
-	 * needed for a secondary cpu after a cold reset e.g
-	 * mark the cpu's presence, mechanism to place it in a
-	 * holding pen etc.
+	 * needed for a CPU to be put into holding pen to wait
+	 * for a warm boot request.
 	 * -----------------------------------------------------
 	 */
-func plat_secondary_cold_boot_setup
+func plat_wait_for_warm_boot
 	/* Calculate address of our hold entry */
 	bl	plat_my_core_pos
 	lsl	x0, x0, #3
 	mov_imm	x2, PLAT_RPI3_TM_HOLD_BASE
 	add	x0, x0, x2
-
 	/*
 	 * This code runs way before requesting the warmboot of this core,
 	 * so it is possible to clear the mailbox before getting a request
 	 * to boot.
 	 */
 	mov	x1, PLAT_RPI3_TM_HOLD_STATE_WAIT
 	str	x1,[x0]
-
-	/* Wait until we have a go */
+	/*
+	 * Wait until we have a go - the sevl deals with the race
+	 * of someone setting a GO state before our wfe.
+	 */
+	sevl
```

Comment:
That would mean just more branches, so I don't think it would be that much cleaner.

Reply:
e.g. there is no wfne is AArch64

Reply:
May need a dsb ld before the ldr x1, [x0]

Reply:
(to match barriers in pwr_domain_on and pwr_down_wfi)

and of course ldr and not ldxr... we're running with no MMU and thus no caches (and thus no expectation that global monitor would work)

Reply:
What I was after is that triggering the local event register just to be able to use a pre-test loop sounds a bit like a hack - if the only reason we do so is to avoid an extra branch.
Regarding the dsb: I don't see why? I think the sev/wfe pair provides a barrier pair already. For the str/ldr sequence in here: this is a proper data dependency, the CPU would do the right thing already.